### PR TITLE
[ObjectsTable] Allow sending selection customization from application

### PR DESCRIPTION
**Pull request by SferaDev**
### 🎯 Objectives

- Allow excluding some rows from parent-children rows

### :memo: Implementation

- Change selection type from ``string[]`` to ``TableSelection[]``
- ``TableSelection`` allows to provide ``checked``, ``indetermindate`` and ``icon`` props to the row CheckBox API.
- Remove old ad-hoc children auto-selection

### :art: Screenshots

image

### CODE REVIEW

File: src/data-table/utils/selection.tsx [hunk 1]
```diff
@@ -1,39 +1,46 @@
...
 }
 
-export function getActionRows<T extends ReferenceObject>(row: T, rows: T[], selection: string[]) {
-    const rowInSelection = !!selection.find(id => row.id === id);
+export function getActionRows<T extends ReferenceObject>(
+    selectedRow: T,
+    allRows: T[],
+    selection: TableSelection[]
+) {
+    const isRowInSelection = !!_.find(selection, { id: selectedRow.id });
```
Comment (tokland) [inline src/data-table/utils/selection.tsx]: minor: `!!_.find` -> `_.some`. 

That `{ id: selectedRow.id }` matcher is cool, but I am afraid TS does not type-check it.

File: src/data-table/utils/selection.tsx [hunk 2]
```diff
@@ -1,39 +1,46 @@
...
+    const selectedRows = _(selection)
+        .map(({ id }) => _.find(allRows, ["id", id]))
```
Comment (tokland) [inline src/data-table/utils/selection.tsx]: A `find` of N nested in a `map` of M is _O(n*m)_, which can get quite large on real scenarios. Typical solution: process the inner data structure so the predicate is O(1).

File: src/old-objects-table/ObjectsTable.js [hunk 3]
```diff
@@ -315,7 +315,10 @@ class ObjectsTable extends React.Component {
         if (_.isEmpty(dataRows) || disableMultiplePageSelection) return [];
 
         const allSelected = selection.size === pager.total;
-        const selectionInOtherPages = _.difference([...selection], dataRows.map(dr => dr.id));
+        const selectionInOtherPages = _.difference(
```
Comment (tokland) [inline src/old-objects-table/ObjectsTable.js:318]: minor: This gets converted to one line when running "yarn prettify"

**Review by tokland — COMMENTED** [review]
No problems with with project-monitoring, I had just to change the selection array. Some comments regarding the code, not very important:

File: src/data-table/utils/selection.tsx (hunk 1 above)
Comment (SferaDev) [reply src/data-table/utils/selection.tsx]: Yeah I noticed that too. I am slowly switching to ``_.find(array, ["id", id])`` but I find the ``_.matchesProperty`` shorthand much more readable.

**Comment by SferaDev** [comment]
@adrianq I will update the PR with these comments during the night. But if you're testing them for metadata-synchronization feel free to merge this as-is and I will add the changes to the next release. Your call!

File: src/data-table/utils/selection.tsx (hunk 1 above)
Comment (SferaDev) [reply src/data-table/utils/selection.tsx]: I did face the type-check error sometimes but here it does indeed properly work :o

image

File: src/data-table/utils/selection.tsx (hunk 2 above)
Comment (SferaDev) [reply src/data-table/utils/selection.tsx]: Simplified with:

```ts
const selectedRows = _.intersectionBy(allRows, selection, "id");
```

Thanks @tokland!

**Review by tokland — APPROVED** [review]
All good!

**Comment by adrianq** [comment]
All good here! Thanks @SferaDev

**Pull request merged**